### PR TITLE
Allow to configure auto logout after browser inactivity

### PR DESCRIPTION
**Pull request by juliusknorr**
With this a a browser will automatically log out the user after no activity during `session_lifetime` even if the `session_keepalive` is enabled.

### CODE REVIEW

File: lib/private/Template/JSConfigHelper.php [hunk 1]
```diff
@@ -176,17 +182,31 @@ public function getConfig() {
 
 		$capabilities = $this->capabilitiesManager->getCapabilities();
 
+		$config = [
```
Comment (skjnldsv) [inline lib/private/Template/JSConfigHelper.php:185]: This file is so awful :see_no_evil: :grin:

Comment (juliushaertl) [reply lib/private/Template/JSConfigHelper.php:185]: The change or the file :stuck_out_tongue_closed_eyes:

Comment (skjnldsv) [reply lib/private/Template/JSConfigHelper.php:185]: The file :stuck_out_tongue:

Comment (juliushaertl) [reply lib/private/Template/JSConfigHelper.php:185]: Ah nevermind, missed the "file", too late already I guess.

**Review by skjnldsv — APPROVED** [review]
Clever approach! Code looks great!

**Review by blizzz — APPROVED** [review]

File: lib/private/Template/JSConfigHelper.php (hunk 1 above)
Comment (rullzer) [reply lib/private/Template/JSConfigHelper.php:185]: We should move a lot of that stuff to the initialstate API I guess. and have proper JS wrappers

Comment (rullzer) [reply lib/private/Template/JSConfigHelper.php:185]: O you do that phew...

**Comment by skjnldsv** [comment]
Jsunit says no

```
"message": "Error: Could not find initial state config of core\nat core/js/dist/main.js:279:1669",
--
997 | "str": "Error: Could not find initial state config of core\nat core/js/dist/main.js:279:1669"
998 | }
999 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.225 secs / 0 secs)
1000 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1001 | {
1002 | "message": "ReferenceError: Can't find variable: OC\nat core/js/tests/specHelper.js:90:3",
1003 | "str": "ReferenceError: Can't find variable: OC\nat core/js/tests/specHelper.js:90:3"
1004 | }
1005 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.227 secs / 0 secs)
1006 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1007 | {
1008 | "message": "ReferenceError: Can't find variable: $\nat core/js/public/publicpage.js:21:3247",
1009 | "str": "ReferenceError: Can't find variable: $\nat core/js/public/publicpage.js:21:3247"
1010 | }
1011 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.228 secs / 0 secs)
1012 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1013 | {
1014 | "message": "ReferenceError: Can't find variable: OC\nat core/js/setupchecks.js:1:31696",
1015 | "str": "ReferenceError: Can't find variable: OC\nat core/js/setupchecks.js:1:31696"
1016 | }
1017 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.228 secs / 0 secs)
1018 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1019 | {
1020 | "message": "ReferenceError: Can't find variable: OCA\nat core/search/js/search.js:21:15042",
1021 | "str": "ReferenceError: Can't find variable: OCA\nat core/search/js/search.js:21:15042"
1022 | }
1023 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.229 secs / 0 secs)
1024 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1025 | {
1026 | "message": "ReferenceError: Can't find variable: OC\nat core/js/mimetype.js:19:8990",
1027 | "str": "ReferenceError: Can't find variable: OC\nat core/js/mimetype.js:19:8990"
1028 | }
1029 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.229 secs / 0 secs)
1030 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1031 | {
1032 | "message": "ReferenceError: Can't find variable: OC\nat core/js/mimetypelist.js:1:646",
1033 | "str": "ReferenceError: Can't find variable: OC\nat core/js/mimetypelist.js:1:646"
1034 | }
1035 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.23 secs / 0 secs)
1036 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1037 | {
1038 | "message": "ReferenceError: Can't find variable: OC\nat core/js/files/fileinfo.js:1:5480",
1039 | "str": "ReferenceError: Can't find variable: OC\nat core/js/files/fileinfo.js:1:5480"
1040 | }
1041 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.23 secs / 0 secs)
1042 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1043 | {
1044 | "message": "ReferenceError: Can't find variable: OC\nat core/js/files/client.js:1:72403",
1045 | "str": "ReferenceError: Can't find variable: OC\nat core/js/files/client.js:1:72403"
1046 | }
1047 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.231 secs / 0 secs)
1048 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1049 | {
1050 | "message": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtags.js:1:4240",
1051 | "str": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtags.js:1:4240"
1052 | }
1053 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.231 secs / 0 secs)
1054 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1055 | {
1056 | "message": "ReferenceError: Can't find variable: Handlebars\nat core/js/systemtags/templates.js:1:29032",
1057 | "str": "ReferenceError: Can't find variable: Handlebars\nat core/js/systemtags/templates.js:1:29032"
1058 | }
1059 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.231 secs / 0 secs)
1060 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1061 | {
1062 | "message": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtagmodel.js:1:3585",
1063 | "str": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtagmodel.js:1:3585"
1064 | }
1065 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.233 secs / 0 secs)
1066 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1067 | {
1068 | "message": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtagscollection.js:1:6710",
1069 | "str": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtagscollection.js:1:6710"
1070 | }
1071 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.233 secs / 0 secs)
1072 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1073 | {
1074 | "message": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtagsmappingcollection.js:1:4799",
1075 | "str": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtagsmappingcollection.js:1:4799"
1076 | }
1077 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.234 secs / 0 secs)
1078 | PhantomJS 2.1.1 (Linux 0.0.0) ERROR
1079 | {
1080 | "message": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtagsinputfield.js:1:35435",
1081 | "str": "ReferenceError: Can't find variable: OC\nat core/js/systemtags/systemtagsinputfield.js:1:35435"
1082 | }
1083 | PhantomJS 2.1.1 (Linux 0.0.0): Executed 0 of 0 ERROR (0.234 secs / 0 secs)
```

**Comment by juliushaertl** [comment]
@ChristophWurst @skjnldsv Any idea how i could properly mock the loadState call in the legacy js tests? 

Otherwise I'd just add a fallback to `OC.config` in https://github.com/nextcloud/server/pull/20298/files#diff-ec0ce8cb7acc57624086f5f4b6a47998R30 as a quick fix :see_no_evil:

**Comment by skjnldsv** [comment]
> @ChristophWurst @skjnldsv Any idea how i could properly mock the loadState call in the legacy js tests?

oh wow :see_no_evil: 
Do the quick fix I guess, Is it later fixable by cleaning?

**Comment by ChristophWurst** [comment]
> @ChristophWurst @skjnldsv Any idea how i could properly mock the loadState call in the legacy js tests?

I'm afraid you can't :cry:

**Comment by skjnldsv** [comment]
I guess this will be for 20 ?

File: config/config.sample.php [hunk 2]
```diff
@@ -251,6 +251,15 @@
...
 'session_keepalive' => true,
 
+/**
+ * Enable or disable the automatic logout after session_lifetime, even if session
+ * keepalive is enabled. This will make sure that an inactive browser will be logged out
+ * even if requests to the server might extend the session lifetime.
+ *
+ * Defaults to ``false``
+ */
+'auto_logout' => false,
```
Comment (rullzer) [inline config/config.sample.php:261]: We should probably make sure that if this is set to true we don't generate a remember me cookie

Comment (juliushaertl) [reply config/config.sample.php:261]: 6052f82a46d24e22db28a9740a970e6700530cf6 should to the trick, please have a look

File: core/src/session-heartbeat.js [hunk 3]
```diff
@@ -83,11 +93,46 @@ const startPolling = () => {
...
+	window.addEventListener('touchstart', e => {
+		lastActive = Date.now()
+		localStorage.setItem('lastActive', lastActive)
+	})
+
+	window.addEventListener('storage', e => {
+		if (e.key !== 'lastActive') {
+			return
+		}
+		lastActive = e.newValue
```
Comment (ChristophWurst) [inline core/src/session-heartbeat.js:116]: lastActive not updated here. is this intentional?

**Review by ChristophWurst — COMMENTED** [review]
Looks good otherwise!

File: core/src/session-heartbeat.js (hunk 3 above)
Comment (juliushaertl) [reply core/src/session-heartbeat.js:116]: Not sure I get your question, it is updated to the value from the local storage.

**Comment by rullzer** [comment]
Guess this is good to go
mind to rebase @juliushaertl then we can do the merge :)

**Comment by juliushaertl** [comment]
So this took me quite some time to figure out why the legacy jsunit tests failed. I pushed a workaround for now (see a143e36 for details), but ideally we should move those tests to the new chai/sinon tests. I'll give that a try in a follow up.

**Comment by skjnldsv** [comment]
> I pushed a workaround for now

It passed

File: core/src/session-heartbeat.js (hunk 3 above)
Comment (ChristophWurst) [reply core/src/session-heartbeat.js:122]: oh no forget my comment. somehow I thought you used storage activity as user activity as well. I didn't get that you use this to communicate across tabs :see_no_evil: 

All good :+1:

**Review by ChristophWurst — APPROVED** [review]

**Comment by skjnldsv** [comment]
So, this recently broke viewer's tests ^^
https://github.com/nextcloud/server/blob/cbd20867b59779d7aff5a024e4287206e3c028e0/lib/private/Template/JSConfigHelper.php#L97-L107
Requires 11 arguments, while 10 are passed here: https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/core/Controller/OCJSController.php#L81-L92

**Comment by skjnldsv** [comment]
Fixing

**Comment by skjnldsv** [comment]
Fix in https://github.com/nextcloud/server/pull/20678

**Pull request merged**